src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -157,18 +158,40 @@ private async Task StartAsync()
                 if (_stream.CanWrite)
                 {
                     sendTask = DoSend();
+                    sendCompletedTask = WaitForWritesCompleted();
                 }
 
                 // Now wait for both to complete
                 await receiveTask;
                 await sendTask;
+                await sendCompletedTask;
+
+                FireStreamClosed();
+
+                await _waitForConnectionClosedTcs.Task;
             }
             catch (Exception ex)
             {
                 _log.LogError(0, ex, $"Unexpected exception in {nameof(QuicStreamContext)}.{nameof(StartAsync)}.");
             }
         }
 
+        private async Task WaitForWritesCompleted()
+        {
+            Debug.Assert(_stream != null);
+
+            try
+            {
+                await _stream.WaitForWriteCompletionAsync();
+                Output.CancelPendingRead();
+            }
+            catch (Exception ex)
+            {
+                // Send error to DoSend loop.
+                Transport.Output.Complete(ex);
```

Comment:
Couldn't middleware still be trying to write to the response at this point? If so, this could lead to InvalidOperationExceptions complaining that "Writing is not allowed after writer was completed."

Why not just call `Output.CancelPendingRead()` in the exceptional case too after storing the exception in a field. I'm not sure, but with so many Exceptions fields already, you might be able to reuse one of them.